sw/device/tests/rstmgr_cpu_info_test.c

```diff
@@ -0,0 +1,202 @@
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+
+/**
+ * RSTMGR CPU INFO TEST
+ *  This test create double fault by accessing register with
+ *  non-existing address.
+ *  After double fault, dut got reset by watch dog bite
+ *  and test collect / check cpu_info from rstmgr
+ */
+
+// non existing address
+#define kIllegalAddr1 0x4041FFF0u
+#define kIllegalAddr2 0x40003618u
+#define kSkipComp 0x12345678u
+#define kCpuDumpSize 9
+
+/**
+ * Cpu dump struct index
+ */
+enum {
+  kCpuDumpIdxCurrentExceptionAddr = 0,
+  kCpuDumpIdxCurrentLastDataAddr = 1,
+  kCpuDumpIdxCurrentNextPc = 2,
+  kCpuDumpIdxCurrentPc = 3,
+  kCpuDumpIdxPreviousExceptionAddr = 4,
+  kCpuDumpIdxPreviousLastDataAddr = 5,
+  kCpuDumpIdxPreviousNextPc = 6,
+  kCpuDumpIdxPreviousPc = 7,
+  kCpuDumpIdxPreviousValid = 8,
+};
+
+/**
+ * Reserve expected cpu dump area in flash
+ */
+__attribute__((section(".non_volatile_scratch")))
+const volatile dif_rstmgr_cpu_info_dump_segment_t exp_dump[kCpuDumpSize] = {
+    UINT32_MAX, UINT32_MAX, UINT32_MAX, UINT32_MAX, UINT32_MAX,
+    UINT32_MAX, UINT32_MAX, UINT32_MAX, UINT32_MAX};
+
+/**
+ *  Dump structure:
+ *    0: current.exception_addr
+ *    1: current.last_data_addr
+ *   *2: current.next_pc
+ *   *3: current.pc
+ *    4: previous.exception_addr
+ *    5: previous.last_data_addr
+ *   *6: previous.next_pc
+ *    7: previous.pc
+ *    8: previous_valid
+ *
+ * Observed cpu dump will be collected after watch dog bite,
+ * exp cpu dump will be created at the 'ottf_exception_handler'.
+ * Following fields are current code specific and
+ * will be skipped comparison.
+ *
+ *  current.next_pc, current.pc, previous.next_pc
+ */
+static dif_rstmgr_cpu_info_dump_segment_t dump[kCpuDumpSize];
+static dif_rstmgr_cpu_info_dump_segment_t temp_dump[kCpuDumpSize] = {
+    kSkipComp, kSkipComp, kSkipComp, kSkipComp, kSkipComp,
+    kSkipComp, kSkipComp, kSkipComp, kSkipComp};
+
+static dif_flash_ctrl_state_t flash_ctrl;
+
+// Count number of falut
+static uint32_t global_error_cnt;
+
+// Access non-existing address
+// Each call will create a fault
+static void read_error(void) {
+  uint32_t addr;
+  global_error_cnt++;
+
+  if (global_error_cnt == 1) {
+    addr = kIllegalAddr1;
+  } else {
+    LOG_INFO("double fault");
+    addr = kIllegalAddr2;
+  }
+  // I can't add a new variable to call mmio_
+  // cause mmio call will never be returned.
+  // Use current variable, just to avoid unused error.
+  addr = mmio_region_read32(mmio_region_from_addr(addr), 0);
+}
+
+/**
+ * Overrides the default OTTF exception handler.
+ *
+ */
+void ottf_exception_handler(void) {
+  uint32_t addr = (uint32_t)exp_dump;
+
+  temp_dump[kCpuDumpIdxCurrentExceptionAddr] =
+      (dif_rstmgr_cpu_info_dump_segment_t)ibex_mepc_read();
+  temp_dump[kCpuDumpIdxPreviousExceptionAddr] =
+      temp_dump[kCpuDumpIdxCurrentExceptionAddr];
+  temp_dump[kCpuDumpIdxCurrentLastDataAddr] = kIllegalAddr2;
+  temp_dump[kCpuDumpIdxPreviousLastDataAddr] = kIllegalAddr2;
```

Comment:
why are these set to the same value? should the previous be `kIllegalAddr1`?

Reply:
That what I thought initially, but from the dump, previous.* field does not snap shot of
previous fault. Can we double check with design owner? 
@rswarbrick  Any thoughts?

Reply:
hmm this is probably on me.
The way the previous works is that on double fault it pushes the current to the previous.
If you look inside `rv_core_ibex`, you'll see the "double fault" logic, is that triggering?

Reply:
![image](https://user-images.githubusercontent.com/99843637/170796371-5f12ede8-d13f-4726-88a7-d0de9c3d25be.png)
yes it is triggered @3343418750ps.
First illegal data address captured @3329508750ps but overwritten by other data
and finally copy 2nd illegal data address

Reply:
@jdonjdon for now, can you create two separate `read_error` functions? one for the first exception and the other for the next? We probably want to make sure that the current/next PC are correctly captured (in addition to mepc). Since the function is the same right now, we can't actually tell.

Reply:
> @jdonjdon for now, can you create two separate `read_error` functions? one for the first exception and the other for the next? We probably want to make sure that the current/next PC are correctly captured (in addition to mepc). Since the function is the same right now, we can't actually tell.

I tried. By doing that, I can see two separate exception_addr are created.
But I can't capture current.exception_addr ( the 2nd fault one), cause it doesn't call
void ottf_exception_handler(void)  in the test.
And last_data_addr doesn't change as we already noticed.

By any chance, do you know how to capture
current exception addr as well as  current/next PC?

Reply:
when you say you can't capture, do you mean the hardware doesn't capture? or that the test doesn't because you don't invoke `ottf_exception_handler` again?

Reply:
not a hardware,
there is overridden ottf_exception_handler in the test file.
it is only called on the first fault.

Reply:
ah yeah that's right. On the second fault cpu execution is forcibly halted by the hardware. 
We probably need to find a way to determine the address without relying on that :)